[Joy] Add `Badge` component

hbjORbj opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/material-ui/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

**This PR includes:**
- Remove unnecessary code in Joy `Avatar` demo
- addition of Joy `Badge` component

**Props**: `color`, `size`, `variant`, `overlap`, `anchorOrigin`, and props of `BadgeUnstyled` component in `@mui/base`

**Default values**:
- `color`: `primary`
- `size`: `md`
- `variant`: `contained`
- `badgeInset`: 0
- `anchorOrigin`: `{vertical: 'top', horizontal: 'right'}`

**Lines of code comparison**
|                          | Material   | Joy        | % change   |
| ------------------------ | ---------- | ---------- | ---------- |
| Badge                    | 401        | 345        | -13.97     |
| total                    | 401        | 345        | -13.97     |

Preview: https://deploy-preview-31401--material-ui.netlify.app/experiments/joy/badge/

danilo-leal commented on packages/mui-joy/src/Badge/Badge.tsx:
```diff
@@ -0,0 +1,326 @@
+        '--Badge-minWidth': '1.25rem',
+        '--Badge-height': '1.25rem',
+        '--Badge-borderRadius': '0.625rem',
+        '--Badge-fontSize': theme.vars.fontSize.md,
```
```suggestion
        '--Badge-fontSize': theme.vars.fontSize.sm,
```
I'd try experimenting with maintaining the small size here, even though we're in the medium-sized badge.

hbjORbj replied:
Sure, done!

danilo-leal commented on packages/mui-joy/src/Badge/Badge.tsx:
```diff
@@ -0,0 +1,326 @@
+      position: 'absolute',
+      boxSizing: 'border-box',
+      fontFamily: theme.vars.fontFamily.body,
+      fontWeight: theme.vars.fontWeight.sm,
```
```suggestion
      fontWeight: theme.vars.fontWeight.md,
```
Would probably iterate with something thicker here...

siriwatknp commented:
**Design**
Can you add try out some of these designs on the experiment page?
- https://dribbble.com/shots/6203221-Rewards-Badges
- https://dribbble.com/shots/3904840-Achievements-UI
- https://www.pinterest.com/pin/482307441353517063/
- https://search.muz.li/ODFhNTgzNTg3
- https://www.pinterest.com/pin/529947081151425115/
- https://www.pinterest.com/pin/703756180567074/

**Accessibility**
The badge in the preview seems like it is clickable but the markup is `span`, I think it is incorrect.

**Props**
- `badgeContent`
  I am not really sure why `<Badge>` has to wrap the children and use `badgeContent` prop to show the badge label. There are cases where using a wrapper can cause the layout to look incorrect (if the Badge is inside a flexbox). Isn't it easier to be customized if it is just a plain `span` that stays inside a relative parent?
  
  ```js
  <IconButton>
    <Badge>content</Badge>
    <Icon />
  </IconButton>
  ```
- `anchorOrigin` remains the same
- `invisible` remains the same
- `max` remains the same
- `overlap` remains the same
- `showZero` remains the same

danilo-leal commented:
> Isn't it easier to be customized if it is just a plain span that stays inside a relative parent? cc @danilo-leal

Seams clearer indeed!

hbjORbj commented:
@siriwatknp 

1. `overlap` prop is added
2. Added a few more examples to the [demo](https://deploy-preview-31401--material-ui.netlify.app/experiments/joy/badge/)
3. Changed `BadgeBadge` component's markup to `button`. Kept `BadgeRoot` to `span`.

danilo-leal commented:
This is looking great, Benny! Great work 👍 

I guess the only thing I'd point out at this point is about the `overlap` prop. By playing around with the circular & rectangular values, I'm not sure what it is actually doing (without inspecting) and if the words are the best options we have 🤔

hbjORbj commented:
> I think by default it should be a span. I don't see that a badge is clickable in [Material UI](https://mui.com/material-ui/react-badge/#accessibility), or do I miss anything?

Oh, I think I misunderstood. On it.

danilo-leal commented:
Would love to get @samuelsycamore's input about the `overlap` prop. Maybe we could find a better term for it?!

samuelsycamore commented:
What exactly does the `overlap` prop do? How would we describe it in documentation?

danilo-leal commented:
The `overlap` prop changes positioning CSS properties. It has a `rectangular` value, that places the badge "outside" of the relying element, and a `circular` element, that places the badge almost in a "perfect" tangent line. Indeed tricky to describe it, I wonder if you'll be able to imagine that only reading. I don't recall seeing this feature in other badge components out there.

@hbjORbj please correct me if I said any non-sense!

samuelsycamore commented:
So if `overlap` is set to `rectangular` then it's not actually "overlapping" the main element? 

Would this be more accurate?

`overlap` => `position`
`rectangular` => `outside`
`circular` => `inside` OR `overlap`

samuelsycamore commented:
What if `overlap` was called `location` instead (keeping the `inside` / `outside` values)? I think it's clear what it means in relation to `anchorOrigin` and avoids confusion with the concept of CSS positioning.

hbjORbj commented:
Sounds good 👍! I will make the change.

danilo-leal reviewed (approved):
Looking great!

siriwatknp commented on packages/mui-joy/src/Badge/Badge.tsx:
```diff
@@ -0,0 +1,387 @@
+      borderRadius: 'var(--Badge-borderRadius)',
+      zIndex: 1, // Render the badge on top of potential ripples.
+      transition: 'transform 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms',
+      ...(ownerState.anchorOrigin!.vertical === 'top' &&
+        ownerState.anchorOrigin!.horizontal === 'right' &&
+        ownerState.location === 'outside' && {
+          top: 0,
+          right: 0,
+          transform: 'scale(1) translate(50%, -50%)',
+          transformOrigin: '100% 0%',
+          [`&.${badgeClasses.invisible}`]: {
+            transform: 'scale(0) translate(50%, -50%)',
+          },
+        }),
+      ...(ownerState.anchorOrigin!.vertical === 'bottom' &&
+        ownerState.anchorOrigin!.horizontal === 'right' &&
+        ownerState.location === 'outside' && {
+          bottom: 0,
+          right: 0,
+          transform: 'scale(1) translate(50%, 50%)',
+          transformOrigin: '100% 100%',
+          [`&.${badgeClasses.invisible}`]: {
+            transform: 'scale(0) translate(50%, 50%)',
+          },
+        }),
```
I guess there should be a better way to simplify this logic. Let me try it first and if I could not find it, we can move forward with this.

siriwatknp replied:
Found it.

siriwatknp commented:
@hbjORbj I pushed several updates.

- Default props: From various designs, I found that the variant is usually a `contained` but the color varies between usage (so I set the color to `primary` instead)
- Fix variables: Some variables are not needed, eg. the `--Badge-minWidth`, and add more meaningful variables (I think the variables should come up by testing with real-world examples).
  - Add `--Badge-ring` variable to create overlap spacing
  - Add `--Badge-minHeight` variable as a baseline for dot. By default, `<Badge>...</Badge>` shows the dot.
- Simplify the conditions (from 100+ lines to 10 lines)
- Add more examples on the experiment page

> Note: I did not change any functional implementation

siriwatknp commented:
@danilo-leal @samuelsycamore I wonder if we should be more open about the customization. What `location` is doing limits the customization to 2 options, either `inside` (offset by -14%) or `outside`.

There are real-world examples where you have to customize this value because of the type of the element that the badge is wrapping (text, avatar, button).

<img width="272" alt="Screen Shot 2565-04-12 at 23 42 47" src="https://user-images.githubusercontent.com/18292247/163012703-0219a82f-6a75-497c-a977-5953b9ac1894.png">

The `location` prop does not matter here because `0%` or `14%` does not work for this case, so you have to do this:

```js
<Badge sx={{ '& .MuiBadge-badge': { top: '2px', right: '-3px' } }}>
  ...
</Badge>
```

I propose to make the customization more robust by using `badgeOffset` (or it can be `location`) prop with the value to target CSS property instead.

```js
<Badge offset={{ top: '2px', right: '-3px', }}>
  ...
</Badge>

// same as location="inside"
<Badge offset="14%">
  ...
</Badge>
```

siriwatknp reviewed (changes requested):
Let's wait for my proposal first.

samuelsycamore commented:
@siriwatknp I think that's a great way to make this feature more robust. 👍 And with that kind of functionality I think `offset` makes sense for a name. In this case, would it make sense for the developer to be able to enter negative numbers? (e.g. `badgeOffset="-14%"` to send the badge further outside and away from the main component. Or maybe that should be flipped—negative values send it further inside, and positive values move it further outside?)

siriwatknp commented:
> @siriwatknp I think that's a great way to make this feature more robust. 👍 And with that kind of functionality I think `offset` makes sense for a name. In this case, would it make sense for the developer to be able to enter negative numbers? (e.g. `badgeOffset="-14%"` to send the badge further outside and away from the main component. Or maybe that should be flipped—negative values send it further inside, and positive values move it further outside?)

I would follow [CSS `outlineOffset` property](https://developer.mozilla.org/en-US/docs/Web/CSS/outline-offset). A negative value moves it further inside, and a positive value moves it further outside.

siriwatknp commented:
I am very happy with the result! It's definitely a huge improvement from Material UI.

<img width="232" alt="Screen Shot 2565-04-15 at 08 34 31" src="https://user-images.githubusercontent.com/18292247/163503845-94cb0b5e-c379-4ab5-a602-dc782aeefdee.png">

```js
<Badge
  color="success"
  size="sm"
  badgeInset="20%"
  anchorOrigin={{ vertical: 'bottom', horizontal: 'right' }}
  sx={{ px: 1 }}
>
  <Avatar size="sm" src="/static/images/avatar/1.jpg" />
</Badge>
<Badge size="sm" badgeInset="2px -3px">
  View 2
</Badge>
```

siriwatknp reviewed (approved):
👍 Great job!
